### ISSUE TITLE
Local IDP XML file doesn't get automatically updated  from public xml URL

### ISSUE THREAD

**Issue by dmitriim** [catalyst/moodle-auth_saml2#25]
If we use URL as "IdP metadata xml OR public xml URL"  when when IDP data has been updated we don't get changes in locally saved idp xml file.

We can probably run cron task to update this data in back end to keep it up to dated.

**Comment by brendanheywood** [comment]
Notes for self: Should be an option not on by default (we've had this cause issues with other clients auto updating bad metadata and breaking).

We should also implement proper certificate rotation as part of this too

**Comment by sam-moodle** [comment]
Hi @brendanheywood, I'm interested in this functionality as well. Did you have an implementation planned? I was thinking this would make most sense as a Moodle scheduled task, but I think that the simplesamlphp library has some built-in support for something similar. I haven't looked too deeply into what benefits using that might have.

**Comment by brendanheywood** [comment]
It would need to be a moodle specific task as it does extra things, eg caching in the sitedata dir. Or at the very least a moodle task which triggers the internal simplesamlphp cron process. It's not high on my radar, did you want to either sponsor it or look at developing this yourself?

**Comment by sam-moodle** [comment]
I think we'd be interested in developing it ourselves. I'd want to make sure that we had the implementation you were looking for (as long as it also met our requirements) so that you could accept the changes as a pull request. Do you have some time to talk about requirements?

**Comment by brendanheywood** [comment]
My requirements would be:

* [ ] opt in via a new admin setting
* [ ] it honors the certificate locking feature https://github.com/catalyst/moodle-auth_saml2/blob/master/certificatelock.php
* [ ] make a new task which just calls the same code to update. Do it in a temp file somewhere else and make sure it passes the same validation that the gui runs before swapping it in (refactor that as needed).
* [ ] as much as feasible is covered by tests
* [ ] all the travis code checkers still run

All up I don't think it will actually be that much code. Ideally I would also have liked to implement certificate rolling (which SSP supports but would need a little moodle wrapping), ie it if the cert in the idp metadata has changed it will support both certs for a month (configurable) and then delete the old one. That may be out of scope for you but would be a nice-to-have.

**Comment by sam-moodle** [comment]
Thanks for that. It's possible I'm being dense, but I don't see the connection between the certificate locking and pulling down the new IDP metadata. Same with the certificate rolling (which I'm not sure we'd be able to bring into scope at this time).

**Comment by brendanheywood** [comment]
Ah sorry yes I got muddled - the certificate is locking the local sp cert, so you are right that's independent. I've removed those from my checklist above

**Comment by sam-moodle** [comment]
Sounds good. When I have an initial PR ready for review I'll let you know.

**Comment by brendanheywood** [comment]
@sam-moodle did you get far with the PR code?

**Comment by brendanheywood** [comment]
@sam-moodle just an fyi that we will probably start working on this in the next few days. So if you have an interim patch then please share it regardless of what state it's in as it could save us a little bit of time. And if not then sit back and wait :)

**Comment by sam-moodle** [comment]
@brendanheywood, unfortunately my time was prioritized with some other things and have not started on the code. Though I did do some research and was going to try to incorporate the meta-refresh module. https://simplesamlphp.org/docs/stable/simplesamlphp-automated_metadata

**Comment by brendanheywood** [comment]
no worries thanks anyway

**Comment by sam-moodle** [comment]
Hey Brendan, could you give me an idea of what your timeline is with this feature? Let me know if I can help in any way still. Code review or writing tests?

**Comment by brendanheywood** [comment]
Sorry I can't give you timelines on that, another internal team in NZ (I'm in Aus) was looking into that for another client. @danmarsden can you shed any light on that?

**Comment by danmarsden** [comment]
I don't think anyone is working on this at the moment here either (but we'd really like to see this done too)

**Comment by sam-moodle** [comment]
@brendanheywood, @danmarsden I've created the following PR #96

**Comment by brendanheywood** [comment]
This is an ugly way of capturing this, but until I've got a bit more time to review and decide which way to go here is an alternate simpler patch:


```diff

From 44f7dfcbf177bc568fe2e03706afdc91595e213c Mon Sep 17 00:00:00 2001
From: Fergus Whyte <fergusw@catalyst.net.nz>
Date: Wed, 8 Feb 2017 17:16:53 +1300
Subject: [PATCH] /auth/saml2/ Added task for Automatic metadatarefresh

WR 257095
---
 auth/saml2/auth.php                        | 91 ++++++++++++++++++++++--------
 auth/saml2/classes/task/fetch_metadata.php | 57 +++++++++++++++++++
 auth/saml2/db/tasks.php                    | 13 +++++
 auth/saml2/lang/en/auth_saml2.php          |  5 +-
 auth/saml2/settings.html                   |  7 +++
 auth/saml2/version.php                     |  4 +-
 6 files changed, 149 insertions(+), 28 deletions(-)
 create mode 100644 auth/saml2/classes/task/fetch_metadata.php
 create mode 100644 auth/saml2/db/tasks.php

diff --git a/auth/saml2/auth.php b/auth/saml2/auth.php
index c21ad6a79f0..40e94419768 100644
--- a/auth/saml2/auth.php
+++ b/auth/saml2/auth.php
@@ -50,6 +50,7 @@ class auth_plugin_saml2 extends auth_plugin_base {
         'spmetadatasign'  => true,
         'showidplink'     => true,
         'alterlogout'     => '',
+        'idpmetadatarefresh' => '0',
     );
 
     /**
@@ -474,44 +475,84 @@ class auth_plugin_saml2 extends auth_plugin_base {
         $form->idpdefaultname = $this->defaults['idpdefaultname'];
         try {
             $rawxml = $form->idpmetadata;
-
+             
             // If rawxml looks like a url, then go scrape it first.
-            if (substr($rawxml, 0, 8) == 'https://' ||
-                substr($rawxml, 0, 7) == 'http://') {
+            $http = (substr($rawxml, 0, 7) == 'http://');
+            $https = (substr($rawxml, 0, 8) == 'https://');
+            
+            if($form->idpmetadatarefresh && $http){
+                $err['idpmetadatarefresh'] = get_string('metadatarefresh_httpsonly','auth_saml2');
+                return;
+            } 
+            
+            if ($https || $http) {
+                   
                 $rawxml = @file_get_contents($rawxml);
-
+                
                 if (!$rawxml) {
                     $err['idpmetadata'] = get_string('idpmetadata_badurl', 'auth_saml2');
                     return;
                 }
+            }else if($form->idpmetadatarefresh){
+                $err['idpmetadatarefresh'] = get_string('idpmetadatarefresh_invalid_nourl', 'auth_saml2');
+                return;
             }
-
-            $xml = new SimpleXMLElement($rawxml);
-            $xml->registerXPathNamespace('md',   'urn:oasis:names:tc:SAML:2.0:metadata');
-            $xml->registerXPathNamespace('mdui', 'urn:oasis:names:tc:SAML:metadata:ui');
-
-            // Find all IDPSSODescriptor elements and then work back up to the entityID.
-            $idps = $xml->xpath('//md:EntityDescriptor[//md:IDPSSODescriptor]');
-            if ($idps && isset($idps[0])) {
-                $form->entityid = (string)$idps[0]->attributes('', true)->entityID[0];
-
-                $names = @$idps[0]->xpath('//mdui:DisplayName');
-                if ($names && isset($names[0])) {
-                    $form->idpdefaultname = (string)$names[0];
-                }
-            }
-
-            if (empty($form->entityid)) {
-                $err['idpmetadata'] = get_string('idpmetadata_noentityid', 'auth_saml2');
-            } else {
+            
+            $xmlitems = $this->validate_xml($rawxml, $err);
+            
+            if(!empty($xmlitems)){
+                $form->entityid = $xmlitems['entityid'];
+                if(isset($xmlitems['idpdefaultname'])) {
+                     $form->idpdefaultname = $xmlitems['idpdefaultname'];
+                }    
                 if (!file_exists($saml2auth->certdir)) {
                     mkdir($saml2auth->certdir);
                 }
                 file_put_contents($saml2auth->certdir . 'idp.xml' , $rawxml);
+ 
+            }
+
+            }catch (Exception $e) {
+                $err['idpmetadata'] = get_string('idpmetadata_invalid', 'auth_saml2');
+            }
+    }
+
+    /* Validates raw xml file and extracts the entityID and Display name. 
+     * 
+     * Returns . 
+     * Ipddefault name is not compulsory. 
+     *@throws XML validation exception.  
+     *@param String Raw xml data.
+     *@param Array of error objects, this will be updated with the error if one is present. 
+     *@return Array containing entityID, and idpdefaultname if present. Null if invalid.
+     */
+    
+
+    public function validate_xml($rawxml, &$err){
+         
+         $xml = new SimpleXMLElement($rawxml);
+         $xml->registerXPathNamespace('md',   'urn:oasis:names:tc:SAML:2.0:metadata');
+         $xml->registerXPathNamespace('mdui', 'urn:oasis:names:tc:SAML:metadata:ui');
+          
+         // Find all IDPSSODescriptor elements and then work back up to the entityID.
+         $idps = $xml->xpath('//md:EntityDescriptor[//md:IDPSSODescriptor]');
+         $xmlvalues = array();
+         if ($idps && isset($idps[0])) {
+            $xmlvalues['entityid'] = (string)$idps[0]->attributes('', true)->entityID[0];
+            $xmlvalues['idpdefaultname'] = null;    
+            $names = @$idps[0]->xpath('//mdui:DisplayName');
+            if ($names && isset($names[0])) {
+              $xmlvalues['idpdefaultname'] = (string)$names[0];
+            }
+            
+            if (empty($xmlvalues['entityid'])) {
+                $err['idpmetadata'] = get_string('idpmetadata_noentityid', 'auth_saml2');
+                return null;
             }
-        } catch (Exception $e) {
-            $err['idpmetadata'] = get_string('idpmetadata_invalid', 'auth_saml2');
         }
+        
+        return $xmlvalues;
+ 
     }
 
     /**
diff --git a/auth/saml2/classes/task/fetch_metadata.php b/auth/saml2/classes/task/fetch_metadata.php
new file mode 100644
index 00000000000..1f21dafe860
--- /dev/null
+++ b/auth/saml2/classes/task/fetch_metadata.php
@@ -0,0 +1,57 @@
+<?php
+namespace auth_saml2\task;
+defined('MOODLE_INTERNAL') || die();
+class fetch_metadata extends \core\task\scheduled_task {
+    public function get_name(){
+    return get_string('fetch_metadata' ,'auth_saml2');
+    }
+
+    public function execute(){
+    
+    $config = get_config('auth/saml2');    
+    
+    if($config->idpmetadatarefresh){
+       $saml2auth = get_auth_plugin('saml2');      
+       if(!substr($config->idpmetadata, 0, 8) == 'https://'){         
+           mtrace("IDP metadata URL not a valid https address");
+           return;
+       }
+       //Let task catch the possible exceptions here. 
+       $newxml = @file_get_contents($config->idpmetadata); 
+       $oldxml = @file_get_contents($saml2auth->certdir . 'idp.xml');
+       
+       if(sha1($newxml) === sha1($oldxml)){
+       //new same as old, end here. 
+           mtrace("Remote metadata identical to local metadata");
+           mtrace("Skipping Task");
+           return;
+       } 
+       
+       $idpvalues = $saml2auth->validate_xml($newxml , $err);
+       
+       if($idpvalues){
+           //set config
+           file_put_contents($saml2auth->certdir . date('U') .'-idp.xml' ,$oldxml);
+           file_put_contents($saml2auth->certdir . 'idp.xml', $newxml);
+       
+           set_config('entityid' , $idpvalues['entityid'] , 'auth_saml2');
+           set_config('idpdefaultname' , $idpvalues['idpdefaultname'] , 'auth_saml2');
+       
+       } else {
+           mtrace("Unable to validate XML at:". $config->ipdmetadata);
+           foreach($err as  $value){
+               mtrace("Error:" . $value);
+           }
+           return;
+       }
+       
+        
+
+    }else{
+       mtrace("IDP metadata refresh disabled: Skipping task" );
+       return; 
+    }
+
+    }
+
+}
diff --git a/auth/saml2/db/tasks.php b/auth/saml2/db/tasks.php
new file mode 100644
index 00000000000..66bacba3d21
--- /dev/null
+++ b/auth/saml2/db/tasks.php
@@ -0,0 +1,13 @@
+<?php
+
+$tasks =array(
+    array(
+        'classname' => 'auth_saml2\task\fetch_metadata',                                                                            
+        'blocking' => 0,                                                                                                            
+        'minute' => '30',                                                                                                            
+        'hour' => '*',                                                                                                              
+        'day' => '*',                                                                                                               
+        'dayofweek' => '*',                                                                                                         
+        'month' => '*'
+    )
+);
diff --git a/auth/saml2/lang/en/auth_saml2.php b/auth/saml2/lang/en/auth_saml2.php
index 98163601ff2..c695ef843ea 100644
--- a/auth/saml2/lang/en/auth_saml2.php
+++ b/auth/saml2/lang/en/auth_saml2.php
@@ -84,6 +84,9 @@ $string['requireint'] = 'This field is required and needs to be a positive integ
 $string['regenerate_submit'] = 'Regenerate';
 $string['test_passive'] = '<a href="{$a}">Test using isPassive</a>';
 $string['test_auth'] = '<a href="{$a}">Test isAuthenticated and login</a>';
-
 $string['alterlogout'] = 'Alternative Logout URL';
 $string['alterlogout_help'] = 'The URL to redirect a user after all internal logout mechanisms are run';
+$string['fetch_metadata'] = 'Retrieve SAML2 IPD metadata from a HTTPS URL';
+$string['idpmetadatarefresh_invalid_nourl'] = 'Metadatarefresh cannot be set without a valid https URL.';
+$string['idpmetadatarefresh'] = 'Enable automatic Idp metadata refresh.';
+$string['idpmetadatarefresh_help'] = 'Automatically retrieve metadata from an Idp URL at a set interval.';
diff --git a/auth/saml2/settings.html b/auth/saml2/settings.html
index 5749c558edb..09d5a1a27c4 100644
--- a/auth/saml2/settings.html
+++ b/auth/saml2/settings.html
@@ -39,6 +39,13 @@ if(!function_exists('mcrypt_encrypt')) {
         <?php print_string($field.'_help', 'auth_saml2') ?></td>
 </tr>
 <tr valign="top">
+    <?php $field = 'idpmetadatarefresh' ?>
+    <td align="right"><label for="<?php echo $field ?>"><?php print_string($field, 'auth_saml2') ?></label></td>
+    <td><?php echo html_writer::select($yesno, $field, $config->$field, false) ?>
+        <?php if (isset($err[$field])) { echo $OUTPUT->notification($err[$field], 'notifyfailure'); } ?>
+        <?php print_string($field.'_help', 'auth_saml2') ?></td>
+</tr>
+<tr valign="top">
     <?php $field = 'idpname' ?>
     <td align="right"><label for="<?php echo $field ?>"><?php print_string($field, 'auth_saml2') ?></label></td>
     <td><input type="text" size="40" name="<?php echo $field ?>" value="<?php print $config->$field ?>" placeholder="<?php print $config->idpdefaultname ?>"><br>
diff --git a/auth/saml2/version.php b/auth/saml2/version.php
index 50039531bb0..94858a8cc5e 100644
--- a/auth/saml2/version.php
+++ b/auth/saml2/version.php
@@ -24,8 +24,8 @@
 
 defined('MOODLE_INTERNAL') || die();
 
-$plugin->version   = 2016090802;    // The current plugin version (Date: YYYYMMDDXX).
-$plugin->release   = 2016090802;    // Match release exactly to version.
+$plugin->version   = 2017021302;    // The current plugin version (Date: YYYYMMDDXX).
+$plugin->release   = 2017021302;   // Match release exactly to version.
 $plugin->requires  = 2014050800;    // Requires this Moodle version.
 $plugin->component = 'auth_saml2';  // Full name of the plugin (used for diagnostics).
 $plugin->maturity  = MATURITY_STABLE;
-- 
2.11.0


```

**Comment by brendanheywood** [comment]
fyi for watchers this is merged in, but only for the 3.3 stable branch. A lot of totaras and older moodles so will keep this open until it's ported

**Comment by roperto** [comment]
The 3.3 branch became master, compatible with older versions.
